### ISSUE TITLE
storeDir does not skip staging of inputs

### ISSUE THREAD

**Issue by anoronh4** [nextflow-io/nextflow#5468]
### Expected behavior and actual behavior

I am expecting storeDir to help skip processes that are already performed. However I am finding that it does not skip staging of the input files when they are remote. If the input file is very large, it will delay the downstream steps of the pipeline to download something that takes up space on the local disk and isn't necessarily used, because the task is skipped.

### Steps to reproduce the problem

```
process A {
cache true
storeDir "storeDir"
input:
tuple val(meta), path(x)
output:
path(x)
script:
"""
echo hi
"""
}

workflow {
remotepath="https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/__genome_libs_StarFv1.10/Mouse_GRCm39_M31_CTAT_lib_Nov092022.source.tar.gz"
A([[:],remotepath])
}

```

### Program output 

If I run for the first time it will download and perform the task, as expected. If I immediately re-run with `-resume` it will skip the process immediately and the pipeline will complete very quickly, also as expected. However, when i remove the folder  `work/stage-*` and then try to resume, I get this:
```
N E X T F L O W  ~  version 23.10.1
Launching `main.nf` [nauseous_archimedes] DSL2 - revision: cc0b1bbdb5
[skipped  ] process > A [100%] 1 of 1, stored: 1 ✔
Staging foreign file: https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/__genome_libs_StarFv1.10/Mouse_GRCm39_M31_CTAT_lib_Nov092022.source.tar.gz
[skipping] Stored process > A
```
This file is only 1.9 Gb which is big enough to notice a delay but it's not very long in the grand scheme of things. But in our pipeline we have a 30 Gb input reference file which takes anywhere from 4-22 minutes to download depending on IO speeds. Is there any way to skip staging? Seems like an unnecessary step in this context.

### Environment 

* Nextflow version: 23.10.1 
* Java version: 11
* Operating system:  Linux
* Bash version: GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)

**Comment by anoronh4** [comment]
I want to add an additional consideration that the pipeline cannot run without access to internet even when the output of a process is in storeDir. Some of our clusters have limited access to internet.

**Comment by bentsherman** [comment]
I think it is because remote file staging happens before checking the storeDir: https://github.com/nextflow-io/nextflow/blob/0a29236e7f070378bc489f149fe3cd4ffd566e31/modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy#L639-L657

Remote file staging happens in `makeTaskContextStage2()`.

I don't see why we couldn't move the remote file staging to after the storeDir check or even later (i.e. right before submitting the task?)

**Comment by bentsherman** [comment]
@jorgee related to #5727, here is another reason why I think we should just move the foreign file staging to right before submitting the task

**Comment by jorgee** [comment]
I will try to move the staging call to the the launch and check if there is not a side effect

**Comment by jorgee** [comment]
Foreign file staging is currently managed in TaskProcessor.makeTaskContextStage2, It is done it two parts: [here](https://github.com/nextflow-io/nextflow/blob/0cf2099043bc1736a81f63827cfd2a03151e60f0/modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy#L2155) it is normalizing the path, checking if the data is in the staging cache and creating the FileCopy if a staging copy is required. It is also setting the resolved path name as input parameter. Later in the same function, it is doing the real copy, that is needed for the hash, because it uses the file attributes to make the hash. 
The good thing is that storeDir is before the hash, so we can modify the makeContextStage2 to return the staging batch and do the copy just before the hash computation. I have created PR #5759 to fix the issue.

**Comment by anoronh4** [comment]
@jorgee yes, this was just a demo. in my actual practice the input is not the same as the output. but if it were, would something like stageInMode 'copy' be enough to get around that? i'm thinking that way whatever is in the stage directory the first time you run it will be fully copied into wherever it is run, and instead the output would be a symlink in the storeDir that points to the copy, not the original. And then hopefully on a resume, nextflow will not try to re-copy or anything like that.

**Comment by jorgee** [comment]
@anoronh4 I have tested with `stageInMode 'copy' `with #5759 and works. It is also working with `stageOutMode 'copy'`. In both cases it is not creating the symlink. In the first case, it copies the data runs the task and move it to the storeDir at the end. In the second case, it uses a symlink during the execution and copies the data at the end.

**Comment by anoronh4** [comment]
thank you, i'll test it out!
